File: Data Structures & Algorithms/invert-a-binary-tree/submission-0.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    // Recursion
    // BC: root is None, return None
    // RC: swap left and right: temp = root.right ; 
    //     root.right = root.left ; root.left = temp
    //      recurse on root.left and root.right
    
    // Time: O(log n). Space: O(log n)
    fun invertTree(root: TreeNode?): TreeNode? {
        if (root == null)
            return null
        
        var rightTemp = root.right
        root.right = invertTree(root.left)
        root.left = invertTree(rightTemp)

        return root
    }
}

File: Data Structures & Algorithms/same-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    # Recursion
    # BC: p = q = None -> return True
    #       p = None and q != None -> False (vice - versa)
    # RC: p = q =!= None
    #     p.val == q.val and recurse on p.left and q.left, same for right

    # Time: O(n). Space: O(h)
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        elif not p or not q:
            return False
        
        if p.val != q.val:
            return False
        else:
            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
        